src/ibex_bluesky_core/callbacks/_kafka.py

```diff
@@ -0,0 +1,80 @@
+import logging
+import os
+import socket
+from typing import Any
+
+import msgpack_numpy
+from bluesky.callbacks import CallbackBase
+from confluent_kafka import Producer
+
+logger = logging.getLogger(__name__)
+
+
+DEFAULT_KAFKA_BROKER = "livedata.isis.cclrc.ac.uk:31092"
+
+
+def get_kafka_topic_name() -> str:
+    """Get the name of the bluesky Kafka topic for this machine."""
+    computer_name = os.environ.get("COMPUTERNAME", socket.gethostname()).upper()
+    computer_name = computer_name.upper()
+    if computer_name.startswith(("NDX", "NDH")):
+        name = computer_name[3:]
+    else:
+        name = computer_name
+
+    return f"{name}_bluesky"
+
+
+class KafkaCallback(CallbackBase):
+    """Forward all bluesky documents to Kafka.
+
+    Documents are sent to Kafka encoded using the MsgPack format with
+    the ``msgpack_numpy`` extension to allow efficiently encoding arrays.
+
+    .. note::
+
+        This callback is automatically configured by
+        :py:obj:`ibex_bluesky_core.run_engine.get_run_engine`, and does not need
+        to be configured manually.
+    """
+
+    def __init__(
+        self,
+        *,
+        bootstrap_servers: list[str] | None = None,
+        topic: str | None = None,
+        key: str,
+        kafka_config: dict[str, Any],
+    ) -> None:
+        super().__init__()
+
+        self._topic = topic or get_kafka_topic_name()
+        self._key = key
+
+        if "bootstrap.servers" in kafka_config:
+            raise ValueError(
+                "Do not specify bootstrap.servers in kafka config, use bootstrap_servers argument."
+            )
+
+        if bootstrap_servers is None:
+            bootstrap_servers = [
+                os.environ.get("IBEX_BLUESKY_CORE_KAFKA_BROKER", DEFAULT_KAFKA_BROKER)
+            ]
+
+        kafka_config["bootstrap.servers"] = ",".join(bootstrap_servers)
+
+        self._producer = Producer(kafka_config)
+
+    def __call__(
```

Comment:
just thinking - we don't want to flush here (to make the producer synchronous) but we could flush the producer in a `__del__()` function.

Reply:
Hmm, I don't think `__del__` works as this callback will always be subsribed (so never deleted).

We _could_ flush after a stop document. Do you think that's better? Still need to test how that would interact with network loss/broker being down/broker being borked.